util/module_logger.go

```diff
@@ -0,0 +1,260 @@
+}
+
+func (ml *ModuleLogger) getParentLogger(*RestrictedLogger) *RestrictedLogger { return ml.l(nil) }
+func (ml *ModuleLogger) Debug(args ...interface{})                           { ml.l(nil).Debug(args...) }
+func (ml *ModuleLogger) Info(args ...interface{})                            { ml.l(nil).Info(args...) }
+func (ml *ModuleLogger) Warn(args ...interface{})                            { ml.l(nil).Warn(args...) }
+func (ml *ModuleLogger) Error(args ...interface{})                           { ml.l(nil).Error(args...) }
+func (ml *ModuleLogger) Fatal(args ...interface{})                           { ml.l(nil).Fatal(args...) }
+func (ml *ModuleLogger) Panic(args ...interface{})                           { ml.l(nil).Panic(args...) }
+func (ml *ModuleLogger) Debugf(format string, args ...interface{})           { ml.l(nil).Debugf(format, args...) }
+func (ml *ModuleLogger) Infof(format string, args ...interface{})            { ml.l(nil).Infof(format, args...) }
+func (ml *ModuleLogger) Warnf(format string, args ...interface{})            { ml.l(nil).Warnf(format, args...) }
+func (ml *ModuleLogger) Errorf(format string, args ...interface{})           { ml.l(nil).Errorf(format, args...) }
+func (ml *ModuleLogger) Fatalf(format string, args ...interface{})           { ml.l(nil).Fatalf(format, args...) }
+func (ml *ModuleLogger) Panicf(format string, args ...interface{})           { ml.l(nil).Panicf(format, args...) }
+func (ml *ModuleLogger) Fields() log.Fields                                  { return ml.l(nil).Fields() }
+
+// SetLogger overrides the logger used for a sub-logger.
+// The root logger, can be referenced using the "root" name.
+func (ml *ModuleLogger) SetLogger(moduleName string, logger log.Logger) {
+	ml.logs[moduleName] = logger
+}
+
+const lowestLevel = DebugLevel
+const highestLevel = OffLevel
+
+// Checking only for greater values is fine as long as Level is unsigned.
+var tooHighErr = fmt.Errorf("minLevel must be less than or equal to %s", highestLevel)
+
+// SetLevel sets the minimum level for a sub-logger. A logger only emits
+// messages that have a level equal to or greater than the one set.
+// Unless configured, a sub-logger outputs all messages.
+func (ml *ModuleLogger) SetLevel(moduleName string, level Level) error {
+	if level > highestLevel {
+		return tooHighErr
+	}
+	ml.levels[moduleName] = level
+	return nil
+}
+
+// Logger returns a sub-logger.
+func (ml *ModuleLogger) Logger(moduleName string) log.Logger {
+	return &ModuleLogger{
+		levels: ml.levels,
+		logs:   ml.logs,
+		l: func(rl *RestrictedLogger) *RestrictedLogger {
+			if rl == nil {
+				l := ml.logs[RootModule]
+				if logger, ok := ml.logs[moduleName]; ok {
+					l = logger
+				}
+				level := lowestLevel
+				if l, ok := ml.levels[moduleName]; ok {
+					level = l
+				}
+				rl = &RestrictedLogger{
+					logger: l,
+					min:    level}
+			}
+			return ml.l(rl)
+		}}
+}
+
+func (ml *ModuleLogger) WithField(key string, value interface{}) log.Logger {
```

Comment:
Yes, sorry, the implementation is a bit opaque, I'll add a comment trying to explain what's going on.

The method is exported because it's part of `log.Logger` interface, and it's basically a wrapper.